dbms/src/IO/HTTPCommon.cpp

```diff
@@ -80,7 +80,7 @@ void setTimeouts(Poco::Net::HTTPClientSession & session, const ConnectionTimeout
 
         ProfileEvents::increment(ProfileEvents::CreatedHTTPConnections);
 
-        session->setHost(DNSResolver::instance().resolveHost(host).toString());
+        session->setHost(host);
```

Comment:
But this will lead to additional DNS resolution (this effectively disables DNS cache).
Instead we should use resolved address as a key in pool.

Reply:
True, but since this is a connection pool, we'll only redial/re-resolve the address when the connection is broken anyway.

I don't think it makes sense to use the DNS cache for connection pools for this reason.

Additionally, if we resolve the address a layer up, then we have a slight leak (since once the IP address changes, we'll never free the connection pool for that address).

You probably have additional context though: Why was the DNS cache added to this section of code to begin with?  It doesn't look like clickhouse hooks into libc, which is how you would ensure that every address resolution is actually going through the cache.

Reply:
> Instead we should use resolved address as a key in pool.

Keying by IP adds one more issue: a mandatory DNS query per each request if we use hostname based URIs. IMO that is basically going to be worse than not caching IP at all. Alternative is to reset cached IP when session is in the bad state (https://github.com/yandex/ClickHouse/pull/5381#issuecomment-495104437).

Reply:
> a mandatory DNS query per each request

How is that the case?  Why would a connection pool execute a DNS query for each request if it's already got connections available for use?

Reply:
> > 
> 
> How is that the case? Why would a connection pool execute a DNS query for each request if it's already got connections available for use?

I mean if key in local session cache is IP, then you'll have to resolve the host before checking if there is a cached session for given URL... But perhaps it is not that bad in presence of DNS cache. 

Anyhow, dandling sessions is still a thing in key-sessions-by-IP approach.

Reply:
Gotcha, that makes sense.  For those who disable the cache entirely it complicates the issue pretty quick.